R/mock2.R

```diff
@@ -0,0 +1,96 @@
+#' Mocking tools
+#'
+#' @description
+#' `r lifecycle::badge("experimental")`
+#'
+#' These functions represent a second attempt at bringing mocking to testthat,
+#' incorporating what we've learned from mockr and mockery There are
+#' two styles of mocking:
+#'
+#' * `local_mocked_bindings()` and `local_mocked_bindings()` work by temporarily
+#'   changing variable bindings in the package namespace. It affects the
+#'   operation of every function in the current package, but only the current
+#'   package.
+#'
+#' * `with_mocked_env()` modifies a function, inserting a new environment in the
+#'   chain of parents with the specified bindings. It can mock any function,
+#'   but because it creates a new function, it doesn't affect any functions
+#'   that use that function.
+#'
+#' @export
+#' @param f A function.
+#' @param ... Name-value pairs providing functions to mock.
+with_mocked_env <- function(f, ...) {
```

Comment:
Is this variant going to be useful? It seems it might be confusing / error prone for users. Even with the note in the documentation, I can see users being confused and still expect that if the mocked function calls another function from the same namespace, then the bindings will be transitively locked, but that won't the case.

Reply:
I originally thought this would cover some cases that `local_mocked_bindings()` would not, but now I don't see where it's useful, unless you want to mock a closure generated by the package, which feels pretty esoteric. Maybe it's fine to drop it?